profiler.go

```diff
@@ -0,0 +1,277 @@
+// If the number is higher than 0, profiler.onTick() will panic after the given number of samples collected.
+var testProfilerPanic int64
+
+func profilerGoroutine(startTime time.Time, result chan<- *profilerResult, stopSignal chan struct{}) {
+	// We shouldn't panic but let's be super safe.
+	defer func() {
+		_ = recover()
+
+		// Make sure we don't block the caller of stopFn() even if we panic.
+		result <- nil
+
+		atomic.StoreInt64(&testProfilerPanic, 0)
+	}()
+
+	// Stop after 30 seconds unless stopped manually.
+	timeout := time.AfterFunc(30*time.Second, func() { stopSignal <- struct{}{} })
+	defer timeout.Stop()
+
+	var localTestProfilerPanic = atomic.LoadInt64(&testProfilerPanic)
+	if localTestProfilerPanic < 0 {
+		panic("This is an expected panic in profilerGoroutine() during tests")
+	}
+
+	profiler := newProfiler(startTime)
+	profiler.testProfilerPanic = localTestProfilerPanic
+
+	// Collect the first sample immediately.
+	profiler.onTick()
+
+	// Periodically collect stacks, starting after profilerSamplingRate has passed.
+	collectTicker := profilerTickerFactory(profilerSamplingRate)
+	defer collectTicker.Stop()
+	var tickerChannel = collectTicker.Channel()
+
+	defer func() {
+		result <- &profilerResult{0, profiler.trace}
+	}()
+
+	for {
+		select {
+		case <-tickerChannel:
+			profiler.onTick()
+		case <-stopSignal:
+			return
+		}
+	}
+}
+
+type profilerResult struct {
+	callerGoID uint64
+	trace      *profileTrace
+}
+
+func getCurrentGoID() uint64 {
+	// We shouldn't panic but let's be super safe.
+	defer func() {
+		_ = recover()
+	}()
+
+	// Buffer to read the stack trace into. We should be good with a small buffer because we only need the first line.
+	var stacksBuffer = make([]byte, 100)
+	var n = runtime.Stack(stacksBuffer, false)
+	if n > 0 {
+		var traces = traceparser.Parse(stacksBuffer[0:n])
+		if traces.Length() > 0 {
+			var trace = traces.Item(0)
+			return trace.GoID()
+		}
+	}
+	return 0
+}
+
+func newProfiler(startTime time.Time) *profileRecorder {
+	// Pre-allocate the profile trace for the currently active number of routines & 100 ms worth of samples.
+	// Other coefficients are just guesses of what might be a good starting point to avoid allocs on short runs.
+	numRoutines := runtime.NumGoroutine()
+	trace := &profileTrace{
+		Frames:         make([]*Frame, 0, 32),
+		Samples:        make([]*profileSample, 0, numRoutines*10), // 100 ms @ 101 Hz
+		Stacks:         make([]profileStack, 0, 8),
+		ThreadMetadata: make(map[string]profileThreadMetadata, numRoutines),
+	}
+
+	return &profileRecorder{
+		startTime:    startTime,
+		trace:        trace,
+		stackIndexes: make(map[string]int, cap(trace.Stacks)),
+		frameIndexes: make(map[string]int, cap(trace.Frames)),
+		// A buffer of 2 KiB per stack looks like a good starting point (empirically determined).
+		stacksBuffer: make([]byte, numRoutines*2048),
+	}
+}
+
+const profilerSamplingRate = time.Second / 101 // 101 Hz
```

Comment:
TIL

```suggestion
const profilerSamplingRate = time.Second / 101 // 101 Hz; not 100 Hz because of the lockstep sampling (https://stackoverflow.com/a/45471031/1181370)
```